Allow value matchers for data and headers in reqs.

ncb000gt opened the pull request:
Requests require exact matching using a `signature` comparison. This restricts the usefulness of mocking since some data may not be possible to get exactly right. In my case I'm sending timestamp data. This provides some flexibility in how strictly the data and headers are matched.

LukaGiorgadze commented:
@ncb000gt This really makes sense.
@GiorgiBeriashvili and @campfire5 already started digging into PR.

Thank you so much!
Happy new year and stay healthy, world needs contributors like you.

ncb000gt commented:
@LukaGiorgadze sounds good, thanks.

In some additional testing I've found a couple issues that I'm working through. I'll push an update once complete.

And you too. :)

ncb000gt commented:
I added a few fixes to the original implementation based on testing I did with my own project. I also added more testing for the changes I've made. Hopefully this is sufficient, but let me know if you have concerns. :) Thanks!

ncb000gt commented:
I rebased `develop` onto this branch and force pushed an update (along with resolving the conflict in the imports).

GiorgiBeriashvili commented:
@ncb000gt First of all, thank you very much for actively working to improve the package. It means a lot to us.

The changes are very useful, most probably necessary for a package such as this. We were trying to analyze the changes and see what **we** could suggest to change. We are content with the matching logic implementation, so we focused on the areas of usability (for other developers) and code style (based on Dart/in-team style conventions), and we made the following changes in form of a suggestion: [branch](https://github.com/lomsa-dev/http-mock-adapter/tree/feature/matchers) ([comparison](https://github.com/lomsa-dev/http-mock-adapter/compare/23256cab92535793d7ece606fd3b640fbfa5c5ec...feature/matchers))

Since the changes we made are mostly based on usability and style, we thought that it might be more suitable not to waste your time and reformat the code ourselves.

Does the edited version satisfy you? Please tell us what you think!

ncb000gt commented:
@GiorgiBeriashvili I took a look at the comparison. Looked good to me.

I think the only thing that really stuck out to me, and is kinda nitpicky, was in `matchesRequest` there are a handful of `has` variables. I'd probably change those to `matches` such as `matchesQueryParameters` or something similar since it doesn't guarantee that they exist, it's just checking that they are the same.

Otherwise seems good to go. :) Thanks for fixing up the style bits!

GiorgiBeriashvili commented:
> @GiorgiBeriashvili I took a look at the comparison. Looked good to me.
> 
> I think the only thing that really stuck out to me, and is kinda nitpicky, was in `matchesRequest` there are a handful of `has` variables. I'd probably change those to `matches` such as `matchesQueryParameters` or something similar since it doesn't guarantee that they exist, it's just checking that they are the same.
> 
> Otherwise seems good to go. :) Thanks for fixing up the style bits!

Good catch! :+1: 

I [updated them](https://github.com/lomsa-dev/http-mock-adapter/commit/41b84f14bf6ee48215257ac902f8af6384756877), hopefully the way you meant it.

Should we merge the changes?

ncb000gt commented:
@GiorgiBeriashvili the commit I made is in the branch you noted so that's good enough for me. :)

I think it looks good overall and makes sense to merge the branch you noted above. Thanks!

GiorgiBeriashvili commented:
Closing this PR as a result of mutual decision where some changes have been applied to #69
